### PR TITLE
Fix: App start time span no longer created if too long

### PR DESCRIPTION
**Pull request by lucas-zimerman**
## :loudspeaker: Type of change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring

## :bulb: Motivation and Context

Fixes #[3044](https://github.com/getsentry/sentry-react-native/issues/3044) / Long span messing the measures of alerts.

## :green_heart: How did you test it?

froze app start and released after the timeout. Before and after  this change can be seen on the following transactions:

https://sentry-sdks.sentry.io/discover/sentry-react-native:6fbe53e73d9048c194169f1228a1bd55/?field=title&field=event.type&field=project&field=user.display&field=timestamp&field=replayId&homepage=true&id=21705&name=&project=5428561&query=&sort=-timestamp&statsPeriod=1h&topEvents=5&yAxis=count()

https://sentry-sdks.sentry.io/discover/sentry-react-native:c764f2fd956e40349e79434d98bcb4ff/?field=title&field=event.type&field=project&field=user.display&field=timestamp&field=replayId&homepage=true&id=21705&name=&project=5428561&query=&sort=-timestamp&statsPeriod=1h&topEvents=5&yAxis=count()

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [ ] All tests passing
- [ ] No breaking changes

### CODE REVIEW

**Comment by lucas-zimerman** [comment]
@krystofwoldrich the transaction will look different on the cases where this issue was happening, is this considered a break change?

**Comment by krystofwoldrich** [comment]
@lucas-zimerman Yes, they are different, but I believe this should be a **fix**, as over minute start-up span is definitely a bug.

**Review by krystofwoldrich — CHANGES_REQUESTED** [review]
This looks good and removes the app start spans longer than the threshold, but the overall transaction is still the same length as the start span would be included. Which I believe is unwanted.

These two transaction start times should be also inside the threshold condition in the `_addAppStartData` method.

https://github.com/getsentry/sentry-react-native/blob/9fce5368945d79430c72cc697041e7c1b6767f3e/src/js/tracing/reactnativetracing.ts#L360

https://github.com/getsentry/sentry-react-native/blob/9fce5368945d79430c72cc697041e7c1b6767f3e/src/js/tracing/reactnativetracing.ts#L469

File: src/js/tracing/reactnativetracing.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { defaultRequestInstrumentationOptions, instrumentOutgoingRequests } from
 import type { Hub, IdleTransaction, Transaction } from '@sentry/core';
 import { getActiveTransaction, getCurrentHub, startIdleTransaction } from '@sentry/core';
 import type { EventProcessor, Integration, Transaction as TransactionType, TransactionContext } from '@sentry/types';
-import { logger } from '@sentry/utils';
+import { logger, timestampInSeconds } from '@sentry/utils';
```
Comment (krystofwoldrich) [inline src/js/tracing/reactnativetracing.ts:7]:
```suggestion
import { logger } from '@sentry/utils';
```

File: src/js/tracing/reactnativetracing.ts [hunk 2]
```diff
@@ -359,20 +374,31 @@ export class ReactNativeTracing implements Integration {
...
    * Adds app start measurements and starts a child span on a transaction.
    */
-  private _addAppStartData(transaction: IdleTransaction, appStart: NativeAppStartResponse): void {
-    if (!this._appStartFinishTimestamp) {
+  private _addAppStartData(
+    transaction: IdleTransaction,
+    appStart: NativeAppStartResponse,
+    appStartDurationMilliseconds: number | undefined,
+  ): void {
+    if (!appStartDurationMilliseconds) {
```
Comment (krystofwoldrich) [inline src/js/tracing/reactnativetracing.ts:390]:
```suggestion
  ): void {
    const appStartDurationMilliseconds = this._getAppStartDurationMilliseconds(appStart);
    if (!appStartDurationMilliseconds) {
```

File: src/js/tracing/reactnativetracing.ts [hunk 3]
```diff
@@ -359,20 +374,31 @@ export class ReactNativeTracing implements Integration {
...
     }
 
+    // we filter out app start more than 60s.
+    // this could be due to many different reasons.
+    // we've seen app starts with hours, days and even months.
+    if (appStartDurationMilliseconds >= ReactNativeTracing._maxAppStart) {
+      return;
+    }
+
     const appStartTimeSeconds = appStart.appStartTime / 1000;
```
Comment (krystofwoldrich) [inline src/js/tracing/reactnativetracing.ts:403]:
```suggestion

    transaction.startTimestamp = appStartTimeSeconds;

```

File: src/js/tracing/reactnativetracing.ts [hunk 4]
```diff
@@ -462,10 +479,14 @@ export class ReactNativeTracing implements Integration {
...
-        transaction.op = 'ui.load';
+        const appStartDurationMilliseconds = this._getAppStartDurationMilliseconds(this._awaitingAppStartData);
+
+        if (appStartDurationMilliseconds && appStartDurationMilliseconds < ReactNativeTracing._maxAppStart) {
+          transaction.startTimestamp = this._awaitingAppStartData.appStartTime / 1000;
+        }
 
-        this._addAppStartData(transaction, this._awaitingAppStartData);
+        transaction.op = UI_LOAD;
+        this._addAppStartData(transaction, this._awaitingAppStartData, appStartDurationMilliseconds);
```
Comment (krystofwoldrich) [inline src/js/tracing/reactnativetracing.ts:489]:
```suggestion
        transaction.op = UI_LOAD;
        this._addAppStartData(transaction, this._awaitingAppStartData);
```

File: test/tracing/reactnativetracing.test.ts [hunk 5]
```diff
@@ -233,6 +234,42 @@ describe('ReactNativeTracing', () => {
...
+          expect(
+            // @ts-expect-error access private for test
+            transaction.spanRecorder,
+          ).toBeDefined();
+
+          expect(
+            // @ts-expect-error access private for test
+            transaction.spanRecorder.spans.some(span => span.op == APP_SPAN_START_WARM),
+          ).toBe(false);
+        }
```
Comment (krystofwoldrich) [inline test/tracing/reactnativetracing.test.ts:270]:
```suggestion

          expect(
            transaction.startTimestamp
          ).toBeGreaterThanOrEqual(timeOriginMilliseconds/1000);
        }
```

File: src/js/tracing/reactnativetracing.ts [hunk 6]
```diff
@@ -350,7 +361,11 @@ export class ReactNativeTracing implements Integration {
     if (this.options.routingInstrumentation) {
       this._awaitingAppStartData = appStart;
     } else {
-      const appStartTimeSeconds = appStart.appStartTime / 1000;
+      const appStartDurationMilliseconds = this._getAppStartDurationMilliseconds(appStart);
+      const appStartTimeSeconds =
+        appStartDurationMilliseconds && appStartDurationMilliseconds < ReactNativeTracing._maxAppStart
+          ? appStart.appStartTime / 1000
+          : timestampInSeconds();
```
Comment (krystofwoldrich) [inline src/js/tracing/reactnativetracing.ts:369]:
```suggestion
```

File: src/js/tracing/reactnativetracing.ts [hunk 7]
```diff
@@ -359,20 +374,31 @@ export class ReactNativeTracing implements Integration {
       });
 
       if (idleTransaction) {
-        this._addAppStartData(idleTransaction, appStart);
+        this._addAppStartData(idleTransaction, appStart, appStartDurationMilliseconds);
```
Comment (krystofwoldrich) [inline src/js/tracing/reactnativetracing.ts:377]:
```suggestion
        this._addAppStartData(idleTransaction, appStart);
```

**Comment by krystofwoldrich** [comment]
@lucas-zimerman Thanks for the update, now the transactions have the correct start timestamp when the app start time is not added. 👍

I just have a few suggestions on how to avoid checking the duration at 3 different places. I believe we can do use the one check in the `addAppStartData` function.

**Review by krystofwoldrich — APPROVED** [review]

**Pull request merged**